src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,254 @@
+   kernel invokes the callback.
+
+Barbara tries to implement the API using
+[`core::future::Future`](https://doc.rust-lang.org/stable/core/future/trait.Future.html)
+so that the library can be compatible with the async Rust ecosystem. The OS
+kernel does not expose a Future-based interface, so Barbara has to implement
+`Future` by hand rather than using async/await syntax. She starts with a
+skeleton:
+
+```rust
+/// Passes `buffer` to the kernel, and prints it to the console. Returns a
+/// future that returns `buffer` when the print is complete. The caller must
+/// call kernel_ready_for_callbacks() when it is ready for the future to return. 
+fn print_buffer(buffer: &'static mut [u8]) -> PrintFuture {
+    // TODO: Set the callback
+    // TODO: Tell the kernel to print `buffer`
+}
+
+struct PrintFuture;
+
+impl core::future::Future for PrintFuture {
+    type Output = &'static mut [u8];
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        // TODO: Detect when the print is done, retrieve `buffer`, and return
+        // it.
+    }
+}
+```
+
+Note: All error handling is omitted to keep things understandable.
+
+Barbara begins to implement `print_buffer`:
+
+```rust
+fn print_buffer(buffer: &'static mut [u8]) -> PrintFuture {
+    kernel_set_print_callback(callback);
+    kernel_start_print(buffer);
+    PrintFuture {}
+}
+
+// New! The callback the kernel calls.
+extern fn callback() {
+    // TODO: Wake up the currently-waiting PrintFuture.
+}
+```
+
+So far so good. Barbara then works on `poll`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            Poll::Ready(kernel_get_buffer_back())
+        }
+        Poll::Pending
+    }
+```
+
+Of course, there's something missing here. How does the callback wake the
+`PrintFuture`? She needs to store the
+[`Waker`](https://doc.rust-lang.org/stable/core/task/struct.Waker.html)
+somewhere! Barbara puts the `Waker` in a global variable so the callback can
+find it (this is fine because the app is single threaded and callbacks do NOT
+interrupt execution the way Unix signals do):
+
+```rust
+static mut PRINT_WAKER: Option<Waker> = None;
+
+extern fn callback() {
+    if let Some(waker) = unsafe { PRINT_WAKER.as_ref() } {
+        waker.wake_by_ref();
+    }
+}
+```
+
+She then modifies `poll` to set `PRINT_WAKER`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            return Poll::Ready(kernel_get_buffer_back());
+        }
+        unsafe { PRINT_WAKER = Some(cx.waker()); }
+        Poll::Pending
+    }
+```
+
+`PRINT_WAKER` is stored in `.bss`, which occupies space in RAM but not flash. It
+is two words in size. It points to a
+[`RawWakerVTable`](https://doc.rust-lang.org/stable/core/task/struct.RawWakerVTable.html)
+that is provided by the executor. `RawWakerVTable` does not appear to be
```

Comment:
> The whole `RawWaker` mechansim was **especially** designed for supporting no-std environments.

I reworded and reorganized this segment significantly, because I definitely didn't convey my thoughts accurately. `RawWaker` was designed to work in both `alloc` and no-`alloc` environments, and it made compromises to support both. I decided to evaluate the overhead of using `Future` when I noticed the compromises in `RawWaker`'s design in order to see if other compromises cost more.

Ultimately, the 2 extra words in `RawWaker` are not a deal-breaker.